### ISSUE TITLE
Null-safety and type converters

### ISSUE THREAD

**Issue by simolus3** [simolus3/drift#1561]
For type converters, `mapToSql` and `mapToDart` use the nullable variant of the types used to declare the converter. This can be a bit confusing, especially for the most common case where those are both non-nullable.

In a breaking release, the generator should enforce a reasonable nullability of type converters:

- `mapToSql` and `mapToDart` are changed to only accept the matching type parameters. To declare a nullable type converter, users would have to extend `TypeConverter<String?, String?>`.
- Report a generation error when a type-converter with a nullable SQL type is used on a non-nullable column
- Report a generation error when a type-converter with a nullable Dart type is used  with a non-nullable field on an existing Dart data class.

**Comment by westito** [comment]
I think the current `NullAwareTypeConverter` do this job right. It needs a small modification (`S extends Object` -> `S extends Object?`), but it can handle nullable and non-null types too. Type nullability depends on type provided in the generics.
My opinion is removing `NullAwareTypeConverter` and modify `TypeConverter` to work like the NullAware would solve this issue. Also have to remove `null_aware_type_converters` option and make the underlining logic permanent.

**Comment by simolus3** [comment]
Yes! Considering the example from your PR, there are cases where null in SQL is not mapped to null in Dart. I think we should be able to express things like this by default (so make `null_aware_type_converters` permanent).

However, `NullAwareTypeConverter` is for mapping `null` to `null` and non-null values to non-null values. This is also why we have `S extends Object`, `requireMapToSql` really should return a non-nullable value. I mostly want to be able to reason about how a converter is dealing with nulls at compile time.

- If it's a `TypeConverter<T, S?>` with a nullable `S?`, we know that `mapToSql` might return null, so we can report a warning if the column is not marked as nullable in the database schema
- If it's a `TypeConverter<T?, S>` with a nullable `T`, we can warn if the target is non-nullable in Dart.

We're doing the second thing with `null_aware_type_converters`, but not the first one. Reliably introducing the first warning and making `null_aware_type_converters` the default requires type converters not "lying" about them returning null. So

```dart
// bad: We don't know how this converter deals with nulls internally
abstract class TypeConverter<D, S> {
  S? mapToSql(D? value);
  D? mapToDart(S? fromDb);
}

// better: we can reason about nullability at compile-time, it's going to be declared with
// nullable types if it wants to return null
abstract class TypeConverter2<D, S> {
  S mapToSql(D value);
  D mapToDart(S fromDb);
}
```

**Comment by westito** [comment]
>If it's a TypeConverter<T?, S> with a nullable T, we can warn if the target is non-nullable in Dart.

What target? Target type depends on `D` parameter, so it always match. 
`S` column nullability warn is okay. But, would be better to throw an error and abort generation. Compiler may throws `Null check operator used on a null value` or will occur a runtime `Null check operator used on a null value` exception.

**Comment by simolus3** [comment]
Fixed in 06af930941407e16b3a0f2f1bdeb5b7d9038fcb2.
